File: app/modules/planning/domain/config/validator.py
"""Configuration validator for planner configurations.

Validates planner configurations and checks for common issues before use.
"""

from typing import List

from app.modules.planning.domain.config.models import PlannerConfiguration


class ConfigurationValidator:
    """Validates planner configurations and checks for issues."""

    @staticmethod
    def validate(config: PlannerConfiguration) -> List[str]:
        """
        Validate configuration and return list of issues.

        Args:
            config: Configuration to validate

        Returns:
            List of validation errors (empty list if valid)
        """
        errors: List[str] = []

        # Check module existence
        from app.modules.planning.domain.calculations.filters.base import (
            sequence_filter_registry,
        )
        from app.modules.planning.domain.calculations.opportunities.base import (
            opportunity_calculator_registry,
        )
        from app.modules.planning.domain.calculations.patterns.base import (
            pattern_generator_registry,
        )
        from app.modules.planning.domain.calculations.sequences.base import (
            sequence_generator_registry,
        )

        # Validate opportunity calculators
        for calc_name in config.get_enabled_calculators():
            if not opportunity_calculator_registry.get(calc_name):
                errors.append(f"Unknown opportunity calculator: '{calc_name}'")

        # Validate pattern generators
        for pattern_name in config.get_enabled_patterns():
            if not pattern_generator_registry.get(pattern_name):
                errors.append(f"Unknown pattern generator: '{pattern_name}'")

        # Validate sequence generators
        for gen_name in config.get_enabled_generators():
            if not sequence_generator_registry.get(gen_name):
                errors.append(f"Unknown sequence generator: '{gen_name}'")

        # Validate filters
        for filter_name in config.get_enabled_filters():
            if not sequence_filter_registry.get(filter_name):
                errors.append(f"Unknown filter: '{filter_name}'")

        # Validate parameter ranges
        if config.max_depth < 1:
            errors.append("max_depth must be >= 1")

        if config.max_depth > 10:
            errors.append(
                "max_depth > 10 may cause excessive computation time (recommended max: 7)"
            )

        if not 0.0 <= config.priority_threshold <= 1.0:
            errors.append("priority_threshold must be between 0.0 and 1.0")

        if config.max_opportunities_per_category < 0:
            errors.append("max_opportunities_per_category must be >= 0")

        if config.transaction_cost_fixed < 0:
            errors.append("transaction_cost_fixed must be >= 0")

        if not 0.0 <= config.transaction_cost_percent <= 1.0:
            errors.append("transaction_cost_percent must be between 0.0 and 1.0")

        if config.transaction_cost_percent > 0.01:
            errors.append(
                f"transaction_cost_percent = {config.transaction_cost_percent} seems unusually high (>1%)"
            )

        # Validate allow_sell and allow_buy
        if not config.allow_sell and not config.allow_buy:
            errors.append("Both allow_sell and allow_buy cannot be False")

        # Validate diversity settings
        if config.enable_diverse_selection:
            if not 0.0 <= config.diversity_weight <= 1.0:
                errors.append("diversity_weight must be between 0.0 and 1.0")

        return errors

    @staticmethod
    def check_dependencies(config: PlannerConfiguration) -> List[str]:
        """
        Check for module dependency issues.

        Args:
            config: Configuration to check

        Returns:
            List of warnings about missing dependencies or suboptimal configs
        """
        warnings: List[str] = []

        # Check if any modules are enabled at all
        enabled_calculators = config.get_enabled_calculators()
        enabled_patterns = config.get_enabled_patterns()
        enabled_generators = config.get_enabled_generators()

        if not enabled_calculators:
            warnings.append(
                "No opportunity calculators enabled - planner will have no opportunities to work with"
            )

        if not enabled_patterns:
            warnings.append(
                "No pattern generators enabled - planner will have no sequences to evaluate"
            )

        # Warn about specific module dependencies
        if (
            "combinatorial" in enabled_generators
            or "enhanced_combinatorial" in enabled_generators
        ):
            if len(enabled_patterns) < 2:
                warnings.append(
                    "Combinatorial generators work best with multiple patterns enabled (currently: {})".format(
                        len(enabled_patterns)
                    )
                )

        if "correlation_aware" in config.get_enabled_filters():
            warnings.append(
                "correlation_aware filter requires securities with correlation data - "
                "ensure risk model is available"
            )

        if "diversity" in config.get_enabled_filters():
            if not config.enable_diverse_selection:
                warnings.append(
                    "diversity filter is enabled but enable_diverse_selection=False - "
                    "filter may have no effect"
                )

        # Check for rebalance patterns without rebalance calculators
        rebalance_patterns = [p for p in enabled_patterns if "rebalance" in p]
        rebalance_calculators = [
            c for c in enabled_calculators if "rebalance" in c or "weight_based" in c
        ]

        if rebalance_patterns and not rebalance_calculators:
            warnings.append(
                f"Rebalance patterns enabled ({rebalance_patterns}) but no rebalance "
                "calculators enabled - patterns may produce empty sequences"
            )

        # Check for profit_taking pattern without profit_taking calculator
        if (
            "profit_taking" in enabled_patterns
            and "profit_taking" not in enabled_calculators
        ):
            warnings.append(
                "profit_taking pattern enabled but profit_taking calculator disabled - "
                "pattern may produce empty sequences"
            )

        # Check for averaging_down pattern without averaging_down calculator
        if (
            "averaging_down" in enabled_patterns
            and "averaging_down" not in enabled_calculators
        ):
            warnings.append(
                "averaging_down pattern enabled but averaging_down calculator disabled - "
                "pattern may produce empty sequences"
            )

        # Warn about performance implications
        if config.max_depth >= 7 and len(enabled_patterns) >= 10:
            warnings.append(
                f"High computational load: max_depth={config.max_depth} with "
                f"{len(enabled_patterns)} patterns may take >60s on Arduino"
            )

        if "combinatorial" in enabled_generators:
            combinatorial_params = config.get_generator_params("combinatorial")
            max_combinations = combinatorial_params.get("max_combinations", 50)
            if max_combinations > 100:
                warnings.append(
                    f"combinatorial generator max_combinations={max_combinations} "
                    "may cause very long planning times"
                )

        # Check if both allow_sell disabled but sell calculators enabled
        if not config.allow_sell:
            sell_calculators = [
                c
                for c in enabled_calculators
                if c in ["profit_taking", "rebalance_sells", "weight_based"]
            ]
            if sell_calculators:
                warnings.append(
                    f"allow_sell=False but sell calculators enabled: {sell_calculators} - "
                    "these calculators will have no effect"
                )

        # Check if both allow_buy disabled but buy calculators enabled
        if not config.allow_buy:
            buy_calculators = [
                c
                for c in enabled_calculators
                if c in ["averaging_down", "opportunity_buys", "rebalance_buys"]
            ]
            if buy_calculators:
                warnings.append(
                    f"allow_buy=False but buy calculators enabled: {buy_calculators} - "
                    "these calculators will have no effect"
                )

        return warnings

    @staticmethod
    def validate_and_warn(
        config: PlannerConfiguration,
    ) -> tuple[bool, List[str], List[str]]:
        """
        Validate configuration and return errors and warnings.

        Args:
            config: Configuration to validate

        Returns:
            Tuple of (is_valid, errors, warnings)
        """
        errors = ConfigurationValidator.validate(config)
        warnings = ConfigurationValidator.check_dependencies(config)

        is_valid = len(errors) == 0

        return (is_valid, errors, warnings)

File: trader/internal/modules/planning/config/validator.go
package config

import (
	"fmt"
	"strings"

	"github.com/aristath/arduino-trader/internal/modules/planning/domain"
)

// ValidationError represents a configuration validation error.
type ValidationError struct {
	Field   string
	Message string
}

func (e ValidationError) Error() string {
	return fmt.Sprintf("%s: %s", e.Field, e.Message)
}

// ValidationErrors represents multiple validation errors.
type ValidationErrors []ValidationError

func (e ValidationErrors) Error() string {
	var messages []string
	for _, err := range e {
		messages = append(messages, err.Error())
	}
	return strings.Join(messages, "; ")
}

// Validator validates planner configurations.
type Validator struct{}

// NewValidator creates a new configuration validator.
func NewValidator() *Validator {
	return &Validator{}
}

// Validate validates a planner configuration.
// Returns ValidationErrors if the configuration is invalid.
func (v *Validator) Validate(config *domain.PlannerConfiguration) error {
	var errors ValidationErrors

	// Validate basic fields
	if config.Name == "" {
		errors = append(errors, ValidationError{
			Field:   "name",
			Message: "name is required",
		})
	}

	// Validate numeric ranges
	if config.MaxDepth <= 0 {
		errors = append(errors, ValidationError{
			Field:   "max_depth",
			Message: "must be greater than 0",
		})
	}

	if config.MaxDepth > 10 {
		errors = append(errors, ValidationError{
			Field:   "max_depth",
			Message: "must be <= 10 (higher values can cause performance issues)",
		})
	}

	if config.MaxOpportunitiesPerCategory <= 0 {
		errors = append(errors, ValidationError{
			Field:   "max_opportunities_per_category",
			Message: "must be greater than 0",
		})
	}

	if config.PriorityThreshold < 0.0 || config.PriorityThreshold > 1.0 {
		errors = append(errors, ValidationError{
			Field:   "priority_threshold",
			Message: "must be between 0.0 and 1.0",
		})
	}

	if config.BeamWidth <= 0 {
		errors = append(errors, ValidationError{
			Field:   "beam_width",
			Message: "must be greater than 0",
		})
	}

	if config.DiversityWeight < 0.0 || config.DiversityWeight > 1.0 {
		errors = append(errors, ValidationError{
			Field:   "diversity_weight",
			Message: "must be between 0.0 and 1.0",
		})
	}

	if config.TransactionCostFixed < 0.0 {
		errors = append(errors, ValidationError{
			Field:   "transaction_cost_fixed",
			Message: "must be >= 0.0",
		})
	}

	if config.TransactionCostPercent < 0.0 {
		errors = append(errors, ValidationError{
			Field:   "transaction_cost_percent",
			Message: "must be >= 0.0",
		})
	}

	// Validate that at least one module is enabled in each category
	enabledCalculators := config.GetEnabledCalculators()
	if len(enabledCalculators) == 0 {
		errors = append(errors, ValidationError{
			Field:   "opportunity_calculators",
			Message: "at least one opportunity calculator must be enabled",
		})
	}

	enabledPatterns := config.GetEnabledPatterns()
	if len(enabledPatterns) == 0 {
		errors = append(errors, ValidationError{
			Field:   "pattern_generators",
			Message: "at least one pattern generator must be enabled",
		})
	}

	enabledGenerators := config.GetEnabledGenerators()
	if len(enabledGenerators) == 0 {
		errors = append(errors, ValidationError{
			Field:   "sequence_generators",
			Message: "at least one sequence generator must be enabled",
		})
	}

	enabledFilters := config.GetEnabledFilters()
	if len(enabledFilters) == 0 {
		errors = append(errors, ValidationError{
			Field:   "filters",
			Message: "at least one filter must be enabled",
		})
	}

	// Validate buy/sell permissions
	if !config.AllowBuy && !config.AllowSell {
		errors = append(errors, ValidationError{
			Field:   "allow_buy/allow_sell",
			Message: "at least one of allow_buy or allow_sell must be true",
		})
	}

	if len(errors) > 0 {
		return errors
	}

	return nil
}

// ValidateQuick performs basic validation without deep checks.
// Useful for quick validation during loading.
func (v *Validator) ValidateQuick(config *domain.PlannerConfiguration) error {
	var errors ValidationErrors

	if config.Name == "" {
		errors = append(errors, ValidationError{
			Field:   "name",
			Message: "name is required",
		})
	}

	if config.MaxDepth <= 0 {
		errors = append(errors, ValidationError{
			Field:   "max_depth",
			Message: "must be greater than 0",
		})
	}

	if len(errors) > 0 {
		return errors
	}

	return nil
}

// ValidateParams validates module-specific parameters.
func (v *Validator) ValidateParams(moduleName string, params map[string]interface{}) error {
	var errors ValidationErrors

	// Define validation rules for common parameter types
	validators := map[string]func(interface{}) error{
		"threshold": func(val interface{}) error {
			if f, ok := val.(float64); ok {
				if f < 0.0 || f > 1.0 {
					return fmt.Errorf("must be between 0.0 and 1.0")
				}
				return nil
			}
			return fmt.Errorf("must be a number")
		},
		"weight": func(val interface{}) error {
			if f, ok := val.(float64); ok {
				if f < 0.0 || f > 1.0 {
					return fmt.Errorf("must be between 0.0 and 1.0")
				}
				return nil
			}
			return fmt.Errorf("must be a number")
		},
		"min_value": func(val interface{}) error {
			if f, ok := val.(float64); ok {
				if f < 0.0 {
					return fmt.Errorf("must be >= 0.0")
				}
				return nil
			}
			return fmt.Errorf("must be a number")
		},
		"max_value": func(val interface{}) error {
			if f, ok := val.(float64); ok {
				if f < 0.0 {
					return fmt.Errorf("must be >= 0.0")
				}
				return nil
			}
			return fmt.Errorf("must be a number")
		},
		"count": func(val interface{}) error {
			if i, ok := val.(int); ok {
				if i <= 0 {
					return fmt.Errorf("must be > 0")
				}
				return nil
			}
			if f, ok := val.(float64); ok {
				if f <= 0 {
					return fmt.Errorf("must be > 0")
				}
				return nil
			}
			return fmt.Errorf("must be an integer")
		},
		"percentage": func(val interface{}) error {
			if f, ok := val.(float64); ok {
				if f < 0.0 || f > 100.0 {
					return fmt.Errorf("must be between 0.0 and 100.0")
				}
				return nil
			}
			return fmt.Errorf("must be a number")
		},
		"factor": func(val interface{}) error {
			if f, ok := val.(float64); ok {
				if f <= 0.0 {
					return fmt.Errorf("must be > 0.0")
				}
				return nil
			}
			return fmt.Errorf("must be a number")
		},
	}

	// Module-specific parameter requirements
	requiredParams := getRequiredParams(moduleName)

	// Check required parameters
	for paramName, paramType := range requiredParams {
		value, exists := params[paramName]
		if !exists {
			errors = append(errors, ValidationError{
				Field:   fmt.Sprintf("%s.%s", moduleName, paramName),
				Message: "required parameter is missing",
			})
			continue
		}

		// Validate parameter type and value
		if validator, ok := validators[paramType]; ok {
			if err := validator(value); err != nil {
				errors = append(errors, ValidationError{
					Field:   fmt.Sprintf("%s.%s", moduleName, paramName),
					Message: err.Error(),
				})
			}
		}
	}

	// Validate cross-parameter constraints
	if err := validateCrossConstraints(moduleName, params); err != nil {
		errors = append(errors, ValidationError{
			Field:   moduleName,
			Message: err.Error(),
		})
	}

	if len(errors) > 0 {
		return errors
	}

	return nil
}

// getRequiredParams returns required parameters for each module type.
func getRequiredParams(moduleName string) map[string]string {
	// Define required parameters per module (paramName -> paramType)
	paramMap := map[string]map[string]string{
		// Opportunity Calculators
		"profit_taking": {
			"gain_threshold": "threshold",
			"windfall_score": "threshold",
			"min_hold_days":  "count",
			"sell_cooldown":  "count",
		},
		"averaging_down": {
			"loss_threshold":    "threshold",
			"max_loss_allowed":  "threshold",
			"buy_cooldown_days": "count",
		},
		"opportunity_buys": {
			"scoring_weight":    "weight",
			"max_opportunities": "count",
		},
		"rebalance_sells": {
			"over_weight_threshold": "threshold",
		},
		"rebalance_buys": {
			"under_weight_threshold": "threshold",
		},
		"weight_based": {
			"target_weight_tolerance": "threshold",
		},

		// Pattern Generators
		"direct_buy":        {},
		"rebalance":         {},
		"single_best":       {},
		"multi_sell":        {},
		"mixed_strategy":    {},
		"opportunity_first": {},
		"deep_rebalance":    {},
		"cash_generation":   {},
		"cost_optimized":    {},
		"adaptive": {
			"adaptation_rate": "factor",
		},
		"market_regime": {
			"regime_threshold": "threshold",
		},

		// Sequence Generators
		"combinatorial": {
			"max_combinations": "count",
		},
		"enhanced_combinatorial": {
			"max_combinations":  "count",
			"pruning_threshold": "threshold",
		},
		"partial_execution": {
			"min_completion_ratio": "threshold",
		},
		"constraint_relaxation": {
			"relaxation_factor": "factor",
		},

		// Filters
		"correlation_aware": {
			"correlation_threshold": "threshold",
		},
		"diversity": {
			"diversity_threshold": "threshold",
		},
		"eligibility": {},
		"recently_traded": {
			"cooldown_days": "count",
		},
	}

	if params, ok := paramMap[moduleName]; ok {
		return params
	}

	// Unknown module - no required params (accept any)
	return map[string]string{}
}

// validateCrossConstraints validates cross-parameter constraints.
func validateCrossConstraints(moduleName string, params map[string]interface{}) error {
	// Check module-specific cross-parameter constraints
	switch moduleName {
	case "profit_taking":
		// Ensure min_hold_days < sell_cooldown
		if minHold, ok1 := params["min_hold_days"].(float64); ok1 {
			if cooldown, ok2 := params["sell_cooldown"].(float64); ok2 {
				if minHold >= cooldown {
					return fmt.Errorf("min_hold_days must be less than sell_cooldown")
				}
			}
		}

	case "averaging_down":
		// Ensure loss_threshold <= max_loss_allowed
		if lossThresh, ok1 := params["loss_threshold"].(float64); ok1 {
			if maxLoss, ok2 := params["max_loss_allowed"].(float64); ok2 {
				if lossThresh > maxLoss {
					return fmt.Errorf("loss_threshold cannot exceed max_loss_allowed")
				}
			}
		}

	case "enhanced_combinatorial":
		// Ensure pruning_threshold is reasonable for max_combinations
		if maxComb, ok1 := params["max_combinations"].(float64); ok1 {
			if pruning, ok2 := params["pruning_threshold"].(float64); ok2 {
				if pruning > 0.9 && maxComb > 1000 {
					return fmt.Errorf("high pruning_threshold with large max_combinations may filter too aggressively")
				}
			}
		}
	}

	return nil
}
